### PR TITLE
Smartbond: Add code to boot MCUboot images

### PR DESCRIPTION
**Pull request by NiekIlmer**
Smartbond: add code to boot MCUboot images

Changes include: updating the cache controller settings on boot in soc.c and updating the EZflashCLI runner to support putting the image in the correct slot

Co-authored-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

Signed-off-by: Niek Ilmer <niek.ilmer.aj@renesas.com>

### CODE REVIEW

File: scripts/west_commands/runners/ezflashcli.py
Comment (nordicjm) [inline scripts/west_commands/runners/ezflashcli.py:46]: Why is there a mcuboot exception?

File: scripts/west_commands/runners/ezflashcli.py
Comment (nordicjm) [inline scripts/west_commands/runners/ezflashcli.py:55]: Can this not flash a hex file instead? I don't know the layout of this chip but if there is a large gap in the memory layout, a hex file is going to be far smaller than a bin file, plus a hex file is transferrable to anywhere and can be programmed, a bin file is basically useless unless you have the zephyr project files.

File: soc/arm/renesas_smartbond/da1469x/soc.c
Comment (nordicjm) [inline soc/arm/renesas_smartbond/da1469x/soc.c:53]: @carlocaione

Comment (carlocaione) [reply soc/arm/renesas_smartbond/da1469x/soc.c:53]: What cache is this and what this code is doing? Asking to check whether this can be hooked up to the cache API instead of having soc-specific code.

Comment (NiekIlmer) [reply soc/arm/renesas_smartbond/da1469x/soc.c:37]: Hi @carlocaione,

The DA1469x has execute in place from external flash, this function configures the cache controller for the XIP mechanism. It also copies the IVT to RAM. Configuring the cache needs a software reset too. I don't think the cache driver is a good fit/ or that other soc's will use this specific mechanism.

Comment (carlocaione) [reply soc/arm/renesas_smartbond/da1469x/soc.c:37]: yeah, I agree, this is fine then.

**Review by andrzej-kaczmarek — DISMISSED** [review]

**Review by nordicjm — CHANGES_REQUESTED** [review]
Marking as -1 until comment questions are answered

File: scripts/west_commands/runners/ezflashcli.py
Comment (andrzej-kaczmarek) [reply scripts/west_commands/runners/ezflashcli.py:58]: ezFlashCLI can only write bin files. It's all written to single location without gaps so file size is not a problem.

**Comment by NiekIlmer** [comment]
Hi @nordicjm 

I hope @andrzej-kaczmarek was able to clear some things up. 

The 'Publish Unit Tests Result' workflow failed, is there something I can do about that?

File: soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series
Comment (andrzej-kaczmarek) [reply soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series:26]: For this SoC it's just more convenient to link at 0x0. Depending on product header, the offset of application image on flash may vary. If application is linked at 0x0 can simply take .bin and write it regardless of product header contents as internal bootloader will remap proper flash offset to 0x0. If using 0x16000000 as a base address we need to know the load offset in advance. This is the case for mcuboot and the reason why boot partition starts at 0x2400 since this is expected load offset with default product header written by ezFlashCLI.

So basically it should be possible to always set `FLASH_BASE_ADDRESS=0x16000000` (or set from dts) but `USE_DT_CODE_PARTITION=n` would also require `FLASH_LOAD_OFFSET=0x2400` to write after "default" product header.

Comment (tejlmand) [reply soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series:26]: other arm based SoC generally uses the value from devicetree, so just for that I believe it would be nice to stick to fetching default value based on devicetree as it will make arm based SoCs more consistent with each other.

And because you will be taking the FLASH_LOAD_OFFSET based on devicetree, but the FLASH_BASE_ADDRESS directly from Kconfig. This gives a slight mismatch wrt. where and how flash and code partition values are defined and obtained.
Mixing flash defaults specied in devicetree with defaults specified directly in Kconfig is always a recipe for confusion.
Ref: https://github.com/zephyrproject-rtos/zephyr/blob/7119cc160531fdd0439c5404cb5d991e44efd98f/Kconfig.zephyr#L100-L103

But afaik there is no official guide that dictate how flash base and flash load offset must be obtained / that it must be obtained  in a specific manner, so I leave the final decision to the maintainer.

File: scripts/west_commands/runners/ezflashcli.py
Comment (tejlmand) [reply scripts/west_commands/runners/ezflashcli.py:46]: @andrzej-kaczmarek it seems you're the original author of this runner, but for others looking at this code, a comment describing what you wrote in above comment would be nice.

**Comment by tejlmand** [comment]
> The 'Publish Unit Tests Result' workflow failed, is there something I can do about that?

Fix the compliance errors (some lines are too long), rebase and push again.
Then hopefully the unit test results will pass.

File: soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series
Comment (andrzej-kaczmarek) [reply soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series]: I updated `FLASH_BASE_ADDRESS` to be set from dts. `FLASH_LOAD_OFFSET` has default value overridden only if `USE_DT_CODE_PARTITION=n` due to mentioned earlier requirements of internal bootloader.

**Review by mbolivar-nordic — COMMENTED** [review]
The runner changes LGTM, but I agree with @tejlmand that a comment describing the get_load_offset() behavior would be an improvement.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
This is no longer a proper patch series; please squash appropriately so we can read the patches in order. Adding get_load_offset() in one patch only to remove it later in the PR is a fixup commit; we don't allow those in zephyr PRs

File: scripts/west_commands/runners/ezflashcli.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/ezflashcli.py:50]: Sorry if I missed some previous discussion in earlier iterations of the patch, but I'm not sure what the value of this symbol is. Since this option is promptless, would it make sense to inline the logic in soc/arm/renesas_smartbond/da1469x/Kconfig.soc into this file?

I.e. write the product header unless `USE_DT_CODE_PARTITION && !MCUBOOT`, which can also be detected from `self.build_conf`, without using the intermediate option. It seems like a lot of SoCs are supported by this tool, so my concern is that this will become a messy list of soc-specific if/else here.

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks for the update! Much easier to read now; I have a question about this new version.

File: scripts/west_commands/runners/ezflashcli.py
Comment (andrzej-kaczmarek) [reply scripts/west_commands/runners/ezflashcli.py]: Yes, that makes sense. I removed option and updated script with proper logic and comments.

**Review by mbolivar-nordic — DISMISSED** [review]
+1 limited to the runner changes only

**Comment by carlescufi** [comment]
@nordicjm please take another look

**Review by andrzej-kaczmarek — DISMISSED** [review]

**Review by sjanc — DISMISSED** [review]

File: boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts
Comment (nordicjm) [inline boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts:83]: How can the size be 0x20000 when the next image starts at 0x8000?

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts
Comment (nordicjm) [reply boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts:83]: In addition to this, unless the scratch partition is really needed because you're transferring between 2 areas with different sector sizes, remove it and increase the size of both slots

**Review by nordicjm — APPROVED** [review]

**Review by andrzej-kaczmarek — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Pull request merged**
